### PR TITLE
17: Add the ability to log in via Google

### PR DESCRIPTION
**Pull request by OlegKarapysh**
Added google authentication flow and shared Google-auth button.
**screenshots ("sign in" and "sign up" texts and alignment will be fixed in task52, they appeared after merge with development branch):**
image
image

### CODE REVIEW

**Comment by OlegKarapysh** [comment]
**screenshots ("sign in" and "sign up" texts and alignment will be fixed in task52, the appeared after merge with development branch):**
image
image

File: backend/Squirrel.Core/Squirrel.Core.Common/DTO/Users/UserDto.cs [hunk 1]
```diff
@@ -1,9 +1,9 @@
 ﻿namespace Squirrel.Core.Common.DTO.Users;
 
-public class UserDTO
+public sealed class UserDto
 {
     public int Id { get; set; }
-    public string Avatar { get; set; }
     public string Email { get; set; }
     public string UserName { get; set; }
+    public string AvatarLink { get; set; }
```
Comment (serhiiyasenev) [inline backend/Squirrel.Core/Squirrel.Core.Common/DTO/Users/UserDto.cs]: `string? AvatarUrl`

**Review by serhiiyasenev — DISMISSED** [review]

File: backend/Squirrel.Core/Squirrel.Core.Common/DTO/Users/UserDto.cs (hunk 1 above)
Comment (OlegKarapysh) [reply backend/Squirrel.Core/Squirrel.Core.Common/DTO/Users/UserDto.cs]: we have 'AvatarLink' property in UserDto.ts. But ok, can be renamed

**Review by serhiiyasenev — DISMISSED** [review]

File: backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        => await _context.Users.FirstOrDefaultAsync(u => u.Username == username);
+
+    public async Task<User> CreateUserAsync(UserRegisterDto userDto, bool isGoogleAuth)
+    {
+        if (await GetUserByUsernameAsync(userDto.Username) is not null)
+        {
+            if (isGoogleAuth)
+            {
+                // Google registration must not fail because of the same username.
+                userDto.Username = GenerateRandomUsername();
```
Comment (Anatolii-Ivanov) [inline backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:36]: We will show userName, so please generate it based on firstname + lastname

Comment (OlegKarapysh) [reply backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:36]: i do this in mapping, but when two users have the same firstname + lastname I generate random unique one

**Review by MaxShelestiuk — DISMISSED** [review]
Everything works and implemented correctly.

**Comment by IlliaLop** [comment]
I agree, it works great)

File: backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs (hunk 2 above)
Comment (alinaSielina) [reply backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:36]: maybe something like this:
firstname + lastname + some random characters

File: backend/Squirrel.Core/Squirrel.Core.BLL/MappingProfiles/UserRegisterProfile.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Squirrel.Core.BLL.MappingProfiles;
+
+public sealed class UserRegisterProfile : Profile
+{
+    public UserRegisterProfile()
+    {
+        CreateMap<User, UserRegisterDto>()!.ReverseMap();
+        CreateMap<GoogleJsonWebSignature.Payload, UserRegisterDto>()
+            .ForMember(m => m.Email, s => s.MapFrom(f => f.Email))
```
Comment (alinaSielina) [inline backend/Squirrel.Core/Squirrel.Core.BLL/MappingProfiles/UserRegisterProfile.cs]: It seems that if the source and destination properties have the same name, AutoMapper will automatically map them by convention even if the mapping is not explicitly defined

Comment (OlegKarapysh) [reply backend/Squirrel.Core/Squirrel.Core.BLL/MappingProfiles/UserRegisterProfile.cs]: yes, I know this, but just wanted to show explicitly what properties I map. Ok, I'll remove it

File: backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs (hunk 2 above)
Comment (OlegKarapysh) [reply backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:36]: and what if this duplicating firstname + lastname takes exactly 25 characters, so we cannot add any random characters to it?)

Comment (alinaSielina) [reply backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:36]: in this case, if we use 5 random characters, we can remove the last 5 characters of the lastname and add 5 random characters. we only want it to be in a more user-friendly format.

**Review by serhiiyasenev — APPROVED** [review]

**Review by alinaSielina — APPROVED** [review]

**Pull request merged**